### PR TITLE
refactor!: Set source link calibrator via Kalman/Ckf options (EDM rework part 1/4)

### PR DESCRIPTION
**Pull request by msmk0**
Move the `calibrator_t` typedef from the Kalman/Ckf types to the respective option types. In general, the calibrator is specific for a given source link implementation and there is no well-defined default behavior. The source link is also only known at call time and might not be defined a construction time. Moving the calibrator to the options means it can be set to match the source link and include information that might only be known at runtime.

The previous default calibrator `VoidMeasurementCalibrator` relied on specific behavior of the `MinimalSourceLink` that is not part of the source link concept. To avoid confusion, the calibrator must now always be set explicitly via the options types and the default implementation has been removed. For use-cases based on the `MinimalSourceLink`, a `MinimalSourceLinkCalibrator` is provided instead.

Additional cleanup: `outlier_finder_t` for the `KalmanFitter` and `source_link_selector_t` for the `CombinatorialKalmanFilter` were previously defined both for the implementation and for the option types. To avoid possible inconsistencies they are now only set via the option types as well.

BREAKING CHANGE: The `KalmanFitter` loses its `calibrator_t` and `outlier_finder_t` template types. The `CombinatorialKalmanFilter` loses its `calibrator_t` and `source_link_selector_t` template types. In both cases, the types are now set via the respective options types and need to be known only at call time instead of already at construction time. The default `VoidMeasurementCalibrator` has been removed.

Includes some additional cleanup/ clarifications in the changed modules.

This is the first PR in what I expect to be a series at least four PRs to rework the measurement event data model types.

### CODE REVIEW

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 1]
```diff
@@ -120,8 +112,11 @@ struct CombinatorialKalmanFilterOptions {
   /// context object for the calibration
   std::reference_wrapper<const CalibrationContext> calibrationContext;
 
+  /// The source link calibrator
+  Calibrator calibrator;
+
   /// The config for the source link selector
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:118]: Not correct anymore.

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 2]
```diff
@@ -945,13 +938,14 @@ class KalmanFitter {
     ACTS_VERBOSE("Preparing " << sourcelinks.size() << " input measurements");
     std::map<const Surface*, source_link_t> inputMeasurements;
     for (const auto& sl : sourcelinks) {
-      const Surface* srf = &sl.referenceSurface();
-      inputMeasurements.emplace(srf, sl);
+      inputMeasurements.emplace(&sl.referenceSurface(), sl);
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:941]: Any reason for this?

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 3]
```diff
@@ -189,24 +185,16 @@ struct KalmanFitterResult {
...
 /// The void components are provided mainly for unit testing.
 template <typename propagator_t, typename updater_t = VoidKalmanUpdater,
-          typename smoother_t = VoidKalmanSmoother,
-          typename outlier_finder_t = VoidOutlierFinder,
-          typename calibrator_t = VoidMeasurementCalibrator>
+          typename smoother_t = VoidKalmanSmoother>
 class KalmanFitter {
- public:
-  /// Shorthand definition
-  using MeasurementSurfaces = std::multimap<const Layer*, const Surface*>;
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:198]: Was this unused?

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 4]
```diff
@@ -71,7 +64,7 @@ struct KalmanFitterOptions {
   KalmanFitterOptions(std::reference_wrapper<const GeometryContext> gctx,
                       std::reference_wrapper<const MagneticFieldContext> mctx,
                       std::reference_wrapper<const CalibrationContext> cctx,
-                      const OutlierFinder& outlierFinder_,
+                      Calibrator&& calibrator_, OutlierFinder&& outlierFinder_,
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:67]: Why not simply by value? They should be trivial structs anyway.

**Review by paulgessinger — COMMENTED** [review]
I think I'm ok with the direction of this PR. Would you be ok to remove the cosmetic changes (maybe also in the follow-up PRs) to make reviewing a bit easier?

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 5]
```diff
@@ -8,7 +8,6 @@
 
 #pragma once
 
-// Workaround for building on clang+libstdc++
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:11]: Why remove this? This include has been moved around frequently, causing issues. This comment indicates not to just that.

File: Core/include/Acts/TrackFinding/CKFSourceLinkSelector.hpp [hunk 6]
```diff
@@ -17,7 +17,6 @@
 #include "Acts/Utilities/TypeTraits.hpp"
 
 #include <limits>
-#include <map>
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CKFSourceLinkSelector.hpp:20]: Unrelated change

File: Core/include/Acts/TrackFinding/CKFSourceLinkSelector.hpp [hunk 7]
```diff
@@ -43,7 +42,6 @@ struct SourceLinkSelectorCuts {
 /// chi2 will be selected and the status will be tagged as an outlier
 ///
 struct CKFSourceLinkSelector {
- public:
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CKFSourceLinkSelector.hpp:46]: Cosmetic change

File: Core/include/Acts/TrackFinding/CKFSourceLinkSelector.hpp [hunk 8]
```diff
@@ -188,7 +186,7 @@ struct CKFSourceLinkSelector {
     return Result<void>::success();
   }
 
-  /// The config
+ private:
   Config m_config;
 };
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CKFSourceLinkSelector.hpp:192]: Cosmetic change

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 9]
```diff
@@ -36,13 +36,12 @@
...
 
 #include <functional>
-#include <map>
 #include <memory>
 #include <unordered_map>
 
 namespace Acts {
 
-/// @brief struct to keep record of the track quality
+/// Track quality summary for one trajectory.
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:44]: Cosmetic change

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 10]
```diff
@@ -60,32 +59,21 @@ struct CombinatorialKalmanFilterTipState {
   size_t nHoles = 0;
 };
 
-/// @brief Options struct how the CombinatorialKalmanFilter (CKF) is called
+/// Combined options for the combinatorial Kalman filter.
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:62]: Cosmetic change

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 11]
```diff
@@ -60,32 +59,21 @@ struct CombinatorialKalmanFilterTipState {
...
 struct CombinatorialKalmanFilterOptions {
-  // Broadcast the source link selector type
+  using Calibrator = calibrator_t;
   using SourceLinkSelector = source_link_selector_t;
 
-  // Broadcast the source link selector config type
-  using SourceLinkSelectorConfig = typename SourceLinkSelector::Config;
-
-  /// Deleted default constructor
-  CombinatorialKalmanFilterOptions() = delete;
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:82]: Cosmetic change

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 12]
```diff
@@ -99,19 +87,23 @@ struct CombinatorialKalmanFilterOptions {
...
+        calibrator(std::move(calibrator_)),
+        sourcelinkSelector(std::move(sourceLinkSelector_)),
         propagatorPlainOptions(pOptions),
         referenceSurface(rSurface),
         multipleScattering(mScattering),
         energyLoss(eLoss),
         smoothing(rSmoothing),
         logger(logger_) {}
+  /// Contexts are required and the options must not be default-constructible.
+  CombinatorialKalmanFilterOptions() = delete;
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:107]: Cosmetic change

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 13]
```diff
@@ -185,16 +180,13 @@ struct CombinatorialKalmanFilterResult {
   Result<void> result{Result<void>::success()};
 };
 
-/// @brief CombinatorialKalmanFilter implementation of Acts as a plugin
+/// Combinatorial Kalman filter to find tracks.
 ///
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:184]: Cosmetic change

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 14]
```diff
@@ -219,24 +211,18 @@ struct CombinatorialKalmanFilterResult {
...
-          typename branch_stopper_t = VoidBranchStopper,
-          typename calibrator_t = VoidMeasurementCalibrator>
+          typename branch_stopper_t = VoidBranchStopper>
 class CombinatorialKalmanFilter {
  public:
-  /// Shorthand definition
-  using MeasurementSurfaces = std::multimap<const Layer*, const Surface*>;
-  /// The navigator type
-  using KalmanNavigator = typename propagator_t::Navigator;
-
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:231]: I have a comment somewhere on `MeasurementSurfaces`, the change to `KalmanNavigator` has nothing to do with the title of this PR.

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 15]
```diff
@@ -219,24 +211,18 @@ struct CombinatorialKalmanFilterResult {
...
   /// Default constructor is deleted
   CombinatorialKalmanFilter() = delete;
-
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:234]: Cosmetic change

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 16]
```diff
@@ -1043,24 +1030,26 @@ class CombinatorialKalmanFilter {
     /// The CombinatorialKalmanFilter smoother
     smoother_t m_smoother;
 
+    /// The measurement calibrator
+    calibrator_t m_calibrator;
+
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:1035]: Cosmetic change

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp [hunk 17]
```diff
@@ -1092,31 +1082,34 @@ class CombinatorialKalmanFilter {
...
   template <typename source_link_container_t, typename start_parameters_t,
+            typename calibrator_t, typename source_link_selector_t,
             typename parameters_t = BoundTrackParameters>
   Result<CombinatorialKalmanFilterResult<
       typename source_link_container_t::value_type>>
   findTracks(const source_link_container_t& sourcelinks,
              const start_parameters_t& sParameters,
-             const CombinatorialKalmanFilterOptions<source_link_selector_t>&
-                 tfOptions) const {
-    const auto& logger = tfOptions.logger;
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:1102]: Cosmetic change

Comment (paulgessinger) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp:1102]: Nevermind, that's the diff.

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp (hunk 5 above)
Comment (paulgessinger) [reply Core/include/Acts/TrackFitting/KalmanFitter.hpp:11]: Cosmetic change

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 18]
```diff
@@ -40,27 +39,21 @@
 
 namespace Acts {
 
-/// @brief Options struct how the Fitter is called
+/// Combined options for the Kalman fitter.
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:42]: Cosmetic change

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 19]
```diff
@@ -80,13 +73,16 @@ struct KalmanFitterOptions {
...
+        calibrator(std::move(calibrator_)),
+        outlierFinder(std::move(outlierFinder_)),
         propagatorPlainOptions(pOptions),
         referenceSurface(rSurface),
         multipleScattering(mScattering),
         energyLoss(eLoss),
         backwardFiltering(bwdFiltering),
         logger(logger_) {}
+  /// Contexts are required and the options must not be default-constructible.
+  KalmanFitterOptions() = delete;
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:85]: Cosmetic change

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 20]
```diff
@@ -157,15 +156,12 @@ struct KalmanFitterResult {
   Result<void> result{Result<void>::success()};
 };
 
-/// @brief Kalman fitter implementation of Acts as a plugin
-///
-/// to the Propgator
+/// Kalman fitter implementation.
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:159]: Cosmetic change

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 21]
```diff
@@ -189,24 +185,16 @@ struct KalmanFitterResult {
...
-  using MeasurementSurfaces = std::multimap<const Layer*, const Surface*>;
   /// The navigator type
   using KalmanNavigator = typename propagator_t::Navigator;
 
   /// The navigator has DirectNavigator type or not
   static constexpr bool isDirectNavigator =
       std::is_same<KalmanNavigator, DirectNavigator>::value;
 
-  /// Default constructor is deleted
-  KalmanFitter() = delete;
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:207]: Cosmetic change

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 22]
```diff
@@ -883,21 +872,23 @@ class KalmanFitter {
     /// The Kalman smoother
     smoother_t m_smoother;
 
+    /// The measurement calibrator
+    calibrator_t m_calibrator;
+
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:877]: Cosmetic change

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 23]
```diff
@@ -883,21 +872,23 @@ class KalmanFitter {
...
     /// The outlier finder
     outlier_finder_t m_outlierFinder;
 
-    /// The Measuremetn calibrator
-    calibrator_t m_calibrator;
-
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:891]: Cosmetic change

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 24]
```diff
@@ -914,9 +905,10 @@ class KalmanFitter {
   /// Fit implementation of the foward filter, calls the
   /// the forward filter and backward smoother
   ///
-  /// @tparam source_link_t Source link type identifying uncalibrated input
-  /// measurements.
+  /// @tparam source_link_t Type of the source link
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:908]: Cosmetic change

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 25]
```diff
@@ -1005,9 +998,10 @@ class KalmanFitter {
...
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:1001]: Same here.

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp [hunk 26]
```diff
@@ -1035,15 +1030,16 @@ class KalmanFitter {
     // To be able to find measurements later, we put them into a map
     // We need to copy input SourceLinks anyways, so the map can own them.
     ACTS_VERBOSE("Preparing " << sourcelinks.size() << " input measurements");
-    std::map<const Surface*, source_link_t> inputMeasurements;
+    std::map<GeometryIdentifier, source_link_t> inputMeasurements;
```
Comment (paulgessinger) [inline Core/include/Acts/TrackFitting/KalmanFitter.hpp:1033]: Oh, is that even correct? Later on you still pass in surface pointers. Are they auto converted? If I'm reading this right, this is dangerous. Also, not the point of this PR, right?

File: Examples/Algorithms/Fitting/src/FittingAlgorithm.cpp [hunk 27]
```diff
@@ -71,7 +71,8 @@ ActsExamples::ProcessCode ActsExamples::FittingAlgorithm::execute(
     const auto& protoTrack = protoTracks[itrack];
     const auto& initialParams = initialParameters[itrack];
 
-    // We can have empty tracks which must give empty fit results
+    // We can have empty tracks which must give empty fit results so the number
+    // of entries in input and output containers matches.
```
Comment (paulgessinger) [inline Examples/Algorithms/Fitting/src/FittingAlgorithm.cpp:75]: Redundant.

File: Examples/Algorithms/Fitting/src/FittingAlgorithm.cpp [hunk 28]
```diff
@@ -113,13 +114,14 @@ ActsExamples::ProcessCode ActsExamples::FittingAlgorithm::execute(
...
-      // Create a SimMultiTrajectory
+      // store the result
       trajectories.emplace_back(std::move(fitOutput.fittedStates),
                                 std::move(trackTips), std::move(indexedParams));
     } else {
       ACTS_WARNING("Fit failed for track " << itrack << " with error"
                                            << result.error());
-      // Fit failed, but still create an empty SimMultiTrajectory
+      // Fit failed. Add an empty result so the output container has
+      // the same number of entries as the input.
```
Comment (paulgessinger) [inline Examples/Algorithms/Fitting/src/FittingAlgorithm.cpp:124]: Arguably better comment, but unrelated to the changeset

File: Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp [hunk 29]
```diff
@@ -19,23 +19,23 @@ ActsExamples::TrackFindingAlgorithm::TrackFindingAlgorithm(
     : ActsExamples::BareAlgorithm("TrackFindingAlgorithm", level),
       m_cfg(std::move(cfg)) {
   if (m_cfg.inputSourceLinks.empty()) {
-    throw std::invalid_argument("Missing input source links collection");
+    throw std::invalid_argument("Missing source links input collection");
```
Comment (paulgessinger) [inline Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp:22]: Cosmetic change

File: Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp [hunk 30]
```diff
@@ -19,23 +19,23 @@ ActsExamples::TrackFindingAlgorithm::TrackFindingAlgorithm(
...
   }
   if (m_cfg.inputInitialTrackParameters.empty()) {
     throw std::invalid_argument(
-        "Missing input initial track parameters collection");
+        "Missing initial track parameters input collection");
```
Comment (paulgessinger) [inline Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp:26]: Cosmetic change

File: Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp [hunk 31]
```diff
@@ -19,23 +19,23 @@ ActsExamples::TrackFindingAlgorithm::TrackFindingAlgorithm(
...
   }
   if (m_cfg.outputTrajectories.empty()) {
-    throw std::invalid_argument("Missing output trajectories collection");
+    throw std::invalid_argument("Missing trajectories output collection");
```
Comment (paulgessinger) [inline Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp:29]: Cosmetic change

File: Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp [hunk 32]
```diff
@@ -50,29 +50,31 @@ ActsExamples::ProcessCode ActsExamples::TrackFindingAlgorithm::execute(
   pOptions.maxSteps = 10000;
 
   // Set the CombinatorialKalmanFilter options
-  ActsExamples::TrackFindingAlgorithm::CKFOptions ckfOptions(
+  ActsExamples::TrackFindingAlgorithm::TrackFinderOptions options(
```
Comment (paulgessinger) [inline Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp:53]:
```suggestion
  ActsExamples::TrackFindingAlgorithm::TrackFinderOptions tfOptions(
```

File: Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithmTrackFinderFunction.cpp [hunk 33]
```diff
@@ -69,10 +69,10 @@ ActsExamples::TrackFindingAlgorithm::makeTrackFinderFunction(
         navigator.resolveMaterial = true;
         navigator.resolveSensitive = true;
         Propagator propagator(std::move(stepper), std::move(navigator));
-        CKF trackFinder(std::move(propagator));
+        Ckf trackFinder(std::move(propagator));
```
Comment (paulgessinger) [inline Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithmTrackFinderFunction.cpp:72]: Cosmetic change, and arbuably, worse. Would you also do `using Kf = KalmanFilter`?

File: Examples/Framework/include/ActsExamples/EventData/SimSourceLink.hpp [hunk 34]
```diff
@@ -44,7 +44,7 @@ class SimSourceLink {
   constexpr const Acts::Surface& referenceSurface() const { return *m_surface; }
   constexpr const ActsFatras::Hit& truthHit() const { return *m_truthHit; }
 
-  Acts::FittableMeasurement<SimSourceLink> operator*() const {
+  Acts::FittableMeasurement<SimSourceLink> makeMeasurement() const {
```
Comment (paulgessinger) [inline Examples/Framework/include/ActsExamples/EventData/SimSourceLink.hpp:47]: Cosmetic change

File: Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp [hunk 35]
```diff
@@ -48,19 +48,33 @@ namespace Test {
...
-  const FittableMeasurement<ExtendedMinimalSourceLink>& operator*() const {
-    return *meas;
+  friend bool operator==(const ExtendedMinimalSourceLink& lhs,
+                         const ExtendedMinimalSourceLink& rhs) {
+    return (lhs.measurement == rhs.measurement);
+  }
+  friend bool operator!=(const ExtendedMinimalSourceLink& lhs,
+                         const ExtendedMinimalSourceLink& rhs) {
+    return (lhs.measurement != rhs.measurement);
+  }
```
Comment (paulgessinger) [inline Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp:69]: Cosmetic change. Also I don't think you need `operator!=`.

**Review by paulgessinger — CHANGES_REQUESTED** [review]
Going through this again, of 16 changes, about 50%+ are purely cosmetic, and/or completely unrelated to the stated objective in this PR. Please rework.

EDIT: I'm happy with 90% of the cosmetic changes (and it's not up to me to judge them). Also, this PR is just about small enough that you can weed through the cosmetic changes, but #511 has 72 (-16 probably), so this is a compounding problem.

File: Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp (hunk 35 above)
Comment (paulgessinger) [reply Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp:69]: Also, (you know my opinion on `friend`, this just changes the operator for the sake of changing it, sorry.

File: Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp [hunk 36]
```diff
@@ -48,19 +48,33 @@ namespace Test {
 
 struct ExtendedMinimalSourceLink {
   size_t sourceID = 0;
+  const FittableMeasurement<ExtendedMinimalSourceLink>* measurement = nullptr;
 
-  const FittableMeasurement<ExtendedMinimalSourceLink>* meas{nullptr};
```
Comment (paulgessinger) [inline Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp:52]: Cosmetic change.

File: Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp (hunk 1 above)
Comment (msmk0) [reply Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp]: Fixed.

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp (hunk 2 above)
Comment (msmk0) [reply Core/include/Acts/TrackFitting/KalmanFitter.hpp]: Left-overs from a change that now only occurs in one of the future PRs. Restored to original implementation.

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp (hunk 3 above)
Comment (msmk0) [reply Core/include/Acts/TrackFitting/KalmanFitter.hpp:198]: Yes, probably a left-over from a previous implementation.

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp (hunk 4 above)
Comment (msmk0) [reply Core/include/Acts/TrackFitting/KalmanFitter.hpp]: Yes, they should be [semiregular](https://en.cppreference.com/w/cpp/concepts/semiregular) since we also copy them by value later on. I switched to by-value usage (also in the CKF options).

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp (hunk 24 above)
Comment (msmk0) [reply Core/include/Acts/TrackFitting/KalmanFitter.hpp:909]: Exactly, that is the point. Says the same thing and fits on the same line.

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp (hunk 26 above)
Comment (msmk0) [reply Core/include/Acts/TrackFitting/KalmanFitter.hpp]: That should of course be `const Surface*`.  This was a left-over from a change that is not part of this PR anymore. I am surprised that this was compiling. Fixed now.

File: Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithmTrackFinderFunction.cpp (hunk 33 above)
Comment (msmk0) [reply Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithmTrackFinderFunction.cpp]: Fixed.

File: Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp (hunk 36 above)
Comment (msmk0) [reply Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp]: Restored.

File: Examples/Framework/include/ActsExamples/EventData/SimSourceLink.hpp (hunk 34 above)
Comment (msmk0) [reply Examples/Framework/include/ActsExamples/EventData/SimSourceLink.hpp:47]: This PR removes the requirement for source links to be dereferenceable. That means there is not need to have this functionality implemented as a special operator overload anymore.

Comment (paulgessinger) [reply Examples/Framework/include/ActsExamples/EventData/SimSourceLink.hpp:47]: But it replaces it with the implicit requirement for there to be `makeMeasurement` method, no?

File: Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp (hunk 35 above)
Comment (msmk0) [reply Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp]: Switched to non-`friend` implementations. The `operator!=` is there since the source link concept will be updated in a future PR to require a [regular](https://en.cppreference.com/w/cpp/concepts/regular) type.

File: Examples/Framework/include/ActsExamples/EventData/SimSourceLink.hpp (hunk 34 above)
Comment (msmk0) [reply Examples/Framework/include/ActsExamples/EventData/SimSourceLink.hpp:47]: No, because the calibrator is not generic, but is usually specific for a given source link type.

File: Examples/Algorithms/Fitting/src/FittingAlgorithm.cpp (hunk 27 above)
Comment (msmk0) [reply Examples/Algorithms/Fitting/src/FittingAlgorithm.cpp:75]: Going through this code, it was not immediately clear why this has to be the case. I needed to check the full loop to understand this. Commented is updated to make it clearer for future readers.

File: Examples/Algorithms/Fitting/src/FittingAlgorithm.cpp (hunk 28 above)
Comment (msmk0) [reply Examples/Algorithms/Fitting/src/FittingAlgorithm.cpp:124]: As above.

File: Core/include/Acts/TrackFitting/KalmanFitter.hpp (hunk 5 above)
Comment (msmk0) [reply Core/include/Acts/TrackFitting/KalmanFitter.hpp]: Fixed.

File: Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp (hunk 35 above)
Comment (paulgessinger) [reply Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp]: Ok maybe I'm dumb, but doesn't the compile synthesize `operator!=()` if you don't supply it explicitly? And `regular` only requires the equality operator as per the link you provided.

**Comment by msmk0** [comment]
@paulgessinger I updated the PR. As discussed privately, I will leave the rest of the comments that point to cosmetic changes unresolved.

File: Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp (hunk 35 above)
Comment (msmk0) [reply Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp]: Not as far as I know.

Comment (paulgessinger) [reply Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp]: I think it does: https://godbolt.org/z/nExK7K

Comment (msmk0) [reply Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp]: I think that is a C++20 feature. See e.g. [cppreference](https://en.cppreference.com/w/cpp/language/operators). If you switch your godbolt example to e.g. gcc 8.3 it does not compile.

Comment (HadrienG2) [reply Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp]: Indeed, just switching the same GCC 10.2 to -std=c++17 breaks this code.

Comment (paulgessinger) [reply Tests/UnitTests/Core/TrackFinding/CombinatorialKalmanFilterTests.cpp]: Hm, I had thought C++20 only added synthetic overloads based on `<=>`, but it does seem like the inequality operator was also added at this point only.

**Review by paulgessinger — APPROVED** [review]
